node/avn-service/src/lib.rs

```diff
@@ -391,6 +397,78 @@ where
     }
 }
 
+#[tokio::main]
+async fn query_token_rates<Block: BlockT, ClientT>(
+    req: tide::Request<Arc<Config<Block, ClientT>>>,
+) -> Result<String, TideError>
+where
+    ClientT: BlockBackend<Block> + UsageProvider<Block> + Send + Sync + 'static,
+{
+    let symbols: Vec<String> = req
+        .param("symbols")
+        .map(|s| s.split(',').map(str::trim).map(String::from).collect())
+        .map_err(|_| TideError::from_str(400, "Missing or invalid symbols parameter"))?;
+
+    let currencies: Vec<String> = req
+        .param("currencies")
+        .map(|s| s.split(',').map(str::trim).map(String::from).collect())
+        .map_err(|_| TideError::from_str(400, "Missing or invalid currency parameter"))?;
+
+    let from: u64 = req
+        .param("from")
+        .map_err(|_| TideError::from_str(400, "Missing from param"))?
+        .parse()
+        .map_err(|_| TideError::from_str(400, "Invalid from param"))?;
+
+    let to: u64 = req
+        .param("to")
+        .map_err(|_| TideError::from_str(400, "Missing to param"))?
+        .parse()
+        .map_err(|_| TideError::from_str(400, "Invalid to param"))?;
+
+    let finance_api_key = &req.state().finance_api_key;
+    let provider =
+        CoinGeckoFinance::new(finance_api_key.clone()).map_err(|e| TideError::from_str(500, e))?;
+
+    let mut results = serde_json::Map::new();
+    for symbol in &symbols {
+        for currency in &currencies {
+            match provider.retrieve_symbol_data(symbol, &currency, from, to).await {
+                Ok(price) => {
+                    log::info!(
+                        "💰 Retrieved price for {} / {} from {} to {}: {}",
+                        symbol,
+                        currency,
+                        from,
+                        to,
+                        price
+                    );
+                    results.insert(
+                        format!("{}", currency),
+                        serde_json::Value::Number(serde_json::Number::from_f64(price).unwrap()),
+                    );
+                },
+                Err(err) => {
+                    log::error!(
```

Comment:
It is a bit odd that we are not returning an error here. We seem to be returning a "valid" response type but with an error string.